starboard/shared/libiamf/iamf_decoder_utils.cc

```diff
@@ -0,0 +1,638 @@
+          // anchor_element and anchored_loudness
+          RCHECK(reader->SkipBytes(3));
+        }
+      }
+      if ((info_type & 0b11111100) > 0) {
+        uint32_t info_type_size;
+        RCHECK(reader->ReadLeb128(&info_type_size));
+        // info_type_bytes
+        RCHECK(reader->SkipBytes(info_type_size));
+      }
+    }
+  }
+
+  // If the mix presentation id is unassigned at this point, the stream is
+  // stereo, or a proper mix presentation for binaural or surround preferred
+  // streams hasn't yet been parsed. Default to the first read mix
+  // presentation in case a preferred mix does not exist.
+  if (!info->mix_presentation_id.has_value()) {
+    info->mix_presentation_id = mix_presentation_id;
+  }
+
+  return true;
+}
+
+bool ParseDescriptorOBU(BufferReader* reader,
+                        IamfBufferInfo* info,
+                        const bool prefer_binaural_audio,
+                        const bool prefer_surround_audio) {
+  SB_DCHECK(reader);
+  uint8_t obu_type = 0;
+  uint32_t obu_size = 0;
+  if (!ParseOBUHeader(reader, &obu_type, &obu_size)) {
+    SB_LOG(ERROR) << "Error parsing OBU header";
+    return false;
+  }
+
+  int next_obu_pos = reader->pos() + obu_size;
+
+  std::optional<uint32_t> binaural_audio_element_id;
+  std::optional<uint32_t> surround_audio_element_id;
+  switch (static_cast<int>(obu_type)) {
+    case kObuTypeCodecConfig:
+      RCHECK(ParseCodecConfigOBU(reader, info));
+      break;
+    case kObuTypeAudioElement:
+      RCHECK(ParseAudioElementOBU(
+          reader, &binaural_audio_element_id, &surround_audio_element_id,
+          prefer_binaural_audio, prefer_surround_audio));
+      break;
+    case kObuTypeSequenceHeader:
+      break;
+    case kObuTypeMixPresentation:
+      RCHECK(ParseMixPresentationOBU(
+          reader, info, &binaural_audio_element_id, &surround_audio_element_id,
+          prefer_binaural_audio, prefer_surround_audio));
+      break;
+    default:
+      // This executes if a non descriptor OBU is read. The buffer info must be
+      // valid by this point.
+      SB_DCHECK(info->is_valid());
+      return info->is_valid();
+  }
+
+  // Skip to the next OBU.
+  const size_t remaining_size = next_obu_pos - reader->pos();
+  RCHECK(reader->SkipBytes(remaining_size));
+  info->config_obus_size = reader->pos();
+  return true;
+}
+
+}  // namespace
+
+bool IamfBufferInfo::is_valid() const {
+  return mix_presentation_id.has_value() && sample_rate > 0 && num_samples > 0;
+}
+
+bool ParseInputBuffer(const scoped_refptr<InputBuffer>& input_buffer,
+                      IamfBufferInfo* info,
+                      const bool prefer_binaural_audio,
+                      const bool prefer_surround_audio) {
+  SB_DCHECK(info);
+  SB_DCHECK(input_buffer->data());
+  SB_DCHECK(!(prefer_binaural_audio && prefer_surround_audio));
+
+  BufferReader reader(input_buffer->data(), input_buffer->size());
+
+  while (!info->is_valid() && reader.pos() < reader.size()) {
```

Comment:
Ideally we should have a fast forward mode that skips all config OBUs so we can return the encoded data.  The detailed config OBU info is only used to configure the decoder once when it's created.

Feel free to add it as a TODO and finish this in future.